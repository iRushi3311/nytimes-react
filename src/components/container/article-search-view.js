import React, { Component } from 'react';
import {connect} from 'react-redux';
import {Pagination} from 'react-bootstrap';
import _ from 'lodash';
import Breadcumb from '../presentational/breadcrumb.js';
import ArticleRow from '../presentational/article-row.js';
import {fetchArticles} from "../../actions/fetch-articles.js";
import AppConstants from '../../constants/app-constants.js';

const mapDispatchToProps = (dispatch) => ({
    fetchArticle: (key) => dispatch(fetchArticles(key))
});

const mapStateToProps = (state) => ({
    articles: state.articlesWrapper.articles,
    totalArticles: state.articlesWrapper.totalArticles,
});

export class ArticleSearchView extends Component {

    constructor(props){
        super(props);
        this.state = {
            articles:[],
            totalArticles: 0,
            activePage: 1,
        };
        this._makeFetchArticleCall = this._makeFetchArticleCall.bind(this);
        this._handlePageSelection = this._handlePageSelection.bind(this);
    }


    componentDidMount(){
        const qs = _.get(this.props, 'match.params.queryString');
        this._makeFetchArticleCall(qs, 0); //On-Load pageOffset is - 0
    }

    componentWillReceiveProps(nextProps){
        const oldQS = _.get(this.props, 'match.params.queryString');
        const newQS = _.get(nextProps, 'match.params.queryString');

        this.setState({
            articles: nextProps.articles,
            totalArticles: nextProps.totalArticles,
        });

        //Reload upon - changed search-Query to view different articles
        if(!_.isEqual(oldQS, newQS)){
            this._makeFetchArticleCall(newQS, 0); //On-Change of Search - pageOffset is - 0
        }
    }

    _makeFetchArticleCall(queryString, nextPage){
        this.props.fetchArticle({key:queryString, pageOffset:nextPage, isSearch:true});
    }

    _handlePageSelection(eventKey){
        const qs = _.get(this.props, 'match.params.queryString');
        let pageOffset = 0;

        this.setState({
            activePage: eventKey
        });

        pageOffset = eventKey - 1 ;
        this._makeFetchArticleCall(qs, pageOffset); //On-Change of Search - pageOffset is - 0

    }

    render() {
        const {backName, currentName} = this.props;
        const {articles, totalArticles} = this.state;
        const totalPages = Math.ceil(totalArticles / AppConstants.PAGE_SIZE);

        return (
            <div>
                <Breadcumb backName={backName} currentName={currentName} />
                {
                    articles.map((article, index) =>
                        <ArticleRow key={index} article={article} />
                    )
                }

                <div style={{textAlign: 'center', marginBottom: '30px'}}>
                    <Pagination prev next first last ellipsis boundaryLinks
                                maxButtons={5} items={totalPages}
                                activePage={this.state.activePage}
                                onSelect={this._handlePageSelection}
                    />
                </div>
            </div>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ArticleSearchView);

