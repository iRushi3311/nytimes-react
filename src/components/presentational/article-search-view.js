import React, { Component } from 'react';
import {connect} from 'react-redux';
import _ from 'lodash';

import Breadcumb from './breadcrumb.js';
import ArticleRow from './article-row.js';

import {fetchArticles} from "../../actions/fetch-articles.js";

const mapDispatchToProps = (dispatch) => ({
    fetchArticle: (key) => dispatch(fetchArticles(key))
});

const mapStateToProps = (state) => ({
    articles: state.articlesWrapper.articles,
    totalArticles: state.articlesWrapper.totalArticles,
});

class ArticleSearchView extends Component {


    constructor(props){
        super(props);

        this.state = {
            articles:[],
            totalArticles: 0,
            currentPage: 0,
            nextPage:0,
        };

        this._makeFetchArticleCall = this._makeFetchArticleCall.bind(this);
    }


    componentDidMount(){
        this._makeFetchArticleCall(this.props);
    }

    componentWillReceiveProps(nextProps){
        const oldQS = _.get(this.props, 'match.params.queryString');
        const newQS = _.get(nextProps, 'match.params.queryString');

        this.setState({
            articles: nextProps.articles,
            totalArticles: nextProps.totalArticles,
        });

        //User changed category to view different articles!
        if(!_.isEqual(oldQS, newQS)){
            this._makeFetchArticleCall(nextProps);
        }
    }

    _makeFetchArticleCall(props){
        debugger;
        const {match} = props;
        const {params} = match;
        const {queryString} = params;
        const {nextPage} = this.state;

        this.props.fetchArticle({key:queryString, pageOffset:nextPage, isSearch:true});
    }

    render() {
        const {backName, currentName, match} = this.props;
        const {articles} = this.state;
        debugger;
        return (
            <div>
                <Breadcumb backName={backName} currentName={currentName} />
                {
                    articles.map((article, index) =>
                        <ArticleRow key={index} article={article} />
                    )
                }
            </div>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ArticleSearchView);

