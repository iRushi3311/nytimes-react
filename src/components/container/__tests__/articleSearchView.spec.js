import React from 'react';
import renderer from 'react-test-renderer';
import {ArticleSearchView} from '../article-search-view.js';
import Article from "../../../models/article";
import {MemoryRouter} from 'react-router-dom';

describe('>>>C O M P O N E N T S --- snapshot test for ArticleSearchView', () => {

    let props,
        componentTree;

    beforeEach(() => {
        const src = {
            _id: 12345,
            web_url: 'google.com',
            headline:{
                main: 'So and so guy did bla bla bla'
            },
            snippet: 'body text body text body text body text body text body text ',
            byline:{
                original: 'By RPatel'
            },
            multimedia: [
                {
                    type: 'type-1',
                    subtype: 'subtype-X',
                    url: 'img_X',
                    height: '75px',
                    width: '75px'
                },
                {
                    type: 'type-3',
                    subtype: 'thumbnail',
                    url: 'img_Z',
                    height: '75px',
                    width: '75px'
                },
            ],
            pub_date: '2017-08-17T09:00:01+0000'
        };
        const articleObj = new Article(src);
        const articles = [articleObj, articleObj, articleObj, articleObj, articleObj, articleObj];

        props = {
            match: {
                params: {
                    queryString: 'obama'
                }
            },
            articles: articles,
            totalArticles: 10,
            fetchArticle: function () {},
        };

        componentTree = renderer.create(
            <MemoryRouter >
                <ArticleSearchView {...props}/>
            </MemoryRouter>
        ).toJSON();
    });

    it('+++ ArticleSearchView match with snapshot ', () => {
        expect(componentTree).toMatchSnapshot();
    });

});