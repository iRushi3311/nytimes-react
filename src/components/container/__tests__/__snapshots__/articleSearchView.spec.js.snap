// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>C O M P O N E N T S --- snapshot test for ArticleSearchView +++ ArticleSearchView match with snapshot  1`] = `
<div>
  <div
    className="sc-bdVaJa bhLhcS"
  >
    <span
      className="sc-bwzfXH gCtkdg"
    >
      <a
        href="/"
        onClick={[Function]}
      >
         
         
      </a>
        
       &gt; 
        
      <a />
    </span>
  </div>
  <div
    style={
      Object {
        "marginBottom": "30px",
        "textAlign": "center",
      }
    }
  >
    <ul
      className="pagination"
    >
      <li
        className="disabled"
        style={undefined}
      >
        <a
          href="#"
          onClick={[Function]}
          role="button"
          style={
            Object {
              "pointerEvents": "none",
            }
          }
          tabIndex={-1}
        >
          <span
            aria-label="First"
          >
            «
          </span>
        </a>
      </li>
      <li
        className="disabled"
        style={undefined}
      >
        <a
          href="#"
          onClick={[Function]}
          role="button"
          style={
            Object {
              "pointerEvents": "none",
            }
          }
          tabIndex={-1}
        >
          <span
            aria-label="Previous"
          >
            ‹
          </span>
        </a>
      </li>
      <li
        className="disabled"
        style={undefined}
      >
        <a
          href="#"
          onClick={[Function]}
          role="button"
          style={
            Object {
              "pointerEvents": "none",
            }
          }
          tabIndex={-1}
        >
          <span
            aria-label="Next"
          >
            ›
          </span>
        </a>
      </li>
      <li
        className="disabled"
        style={undefined}
      >
        <a
          href="#"
          onClick={[Function]}
          role="button"
          style={
            Object {
              "pointerEvents": "none",
            }
          }
          tabIndex={-1}
        >
          <span
            aria-label="Last"
          >
            »
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
